flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/AsyncStateTableStreamOperator.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.runtime.operators;
+
+import org.apache.flink.core.memory.ManagedMemoryUseCase;
+import org.apache.flink.runtime.asyncprocessing.operators.AbstractAsyncStateStreamOperator;
+import org.apache.flink.runtime.execution.Environment;
+import org.apache.flink.streaming.api.TimerService;
+import org.apache.flink.streaming.api.operators.InternalTimerService;
+import org.apache.flink.streaming.api.watermark.Watermark;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+import org.apache.flink.streaming.runtime.tasks.ProcessingTimeService;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+import static org.apache.flink.util.Preconditions.checkState;
+
+/**
+ * Table operator to invoke close always. This is a base class for both batch and stream operators
+ * without key.
+ *
+ * <p>This class is nearly identical with {@link TableStreamOperator}, but extending from {@link
+ * AbstractAsyncStateStreamOperator} to integrate with asynchronous state access.
+ */
+public abstract class AsyncStateTableStreamOperator<OUT>
+        extends AbstractAsyncStateStreamOperator<OUT> {
+
+    /** We listen to this ourselves because we don't have an {@link InternalTimerService}. */
+    protected long currentWatermark = Long.MIN_VALUE;
+
+    protected transient ContextImpl ctx;
+
+    @Override
+    public void open() throws Exception {
+        super.open();
+        this.ctx = new ContextImpl(getProcessingTimeService());
+    }
+
+    @Override
+    public boolean useSplittableTimers() {
+        return true;
+    }
+
+    /** Compute memory size from memory faction. */
+    public long computeMemorySize() {
```

Comment:
When will this calculation be used？

Reply:
I just full copy `TableStreamOperator` as `AsyncStateTableStreamOperator`. I will delete it until it actually is needed.